### ISSUE TITLE
Crash on Esc out of Prefab Edit Mode

### ISSUE THREAD

**Issue by michabr** [o3de/o3de#11741]
**Describe the bug**
When drilling into a nested prefab in Prefab Edit Mode, then pressing Esc, a crash occurs. This doesn't always happen the first time but repeating a second time causes a crash.

**Assets required**
[Assets.zip](https://github.com/o3de/o3de/files/9476873/Assets.zip)

**Steps to reproduce**
1. Extract the attached .prefab files to project folder
2. Open Editor with any level
3. Instantiate "TruckPrefab" prefab
4. Double click the new instance to go into Prefab Edit Mode
5. Double click on the "WheelsPrefab"
6. Double click on the "Wheel" prefab"
7. Press Esc
8. Repeat

**Expected behavior**
No crash.

**Actual behavior**
Crash.

**Screenshots/Video**
If applicable, add screenshots and/or a video to help explain your problem.

**Found in Branch**
Stabilization.

**Desktop/Device (please complete the following information):**
 - Device: [e.g. PC, Mac, iPhone, Samsung] 
 - OS: [e.g. Windows, macOS, iOS, Android]
 - Version [e.g. 10, Monterey, Oreo]
 - CPU [e.g. Intel I9-9900k , Ryzen 5900x, ]
 - GPU [AMD 6800 XT, NVidia RTX 3090]
 - Memory [e.g. 16GB]

**Additional context**

**Comment by forkercat** [comment]
I was able to reproduce it. A pretty consistent crash. The reproduce steps could be simplified as follows:
1. Extract the `Assets.zip` and put `Assets` folder to the project root. (in my case, I put in under `AutomatedTesting/Assets/*.prefab`)
2. Create an empty level
3. Instantiate `TrucksPrefab` (make sure it has all the nested prefabs... otherwise, something is wrong in step 1)
4. Once you're sure all the prefab files are loaded correctly
5. Create an empty level again
6. Instantiate `TrucksPrefab`
7. Double-click the prefab to enter Prefab Edit Mode
8. Hit Esc to exit
9. Double-click the prefab to enter Prefab Edit Mode
10. Hit Esc to exit
11. Crash

<img height="300px" src="https://user-images.githubusercontent.com/11157226/188725448-1f2b9e48-8cb0-4fdf-afb3-fa8abf42e7a1.png">

**Comment by forkercat** [comment]
There are several things I noticed in the template file.

(1) The provided `TruckPrefab.prefab` has incorrect child entity order. It should be three items in the list. But this is not the cause to the crash.
```json
"Component_[9470417891000702028]": {
    "$type": "EditorEntitySortComponent",
    "Id": 9470417891000702028,
    "Child Entity Order": [
        "Instance_[3806333159060]/ContainerEntity",
        "Instance_[4257304725140]/ContainerEntity"
    ]
}
```

(2) The provided asset has multiple nested children, but the crash can be reproduced even for one child like this:
```
Level
    TruckPrefab    (prefab)
        Truck
            WheelsPrefab    (prefab)
                Wheels
                    Wheel1    (prefab)
                        Wheel
```

(3) The wheel prefab has a mesh component attached. Removing the mesh component does not resolve the crash.

Simplified assets with only one child and without mesh component: [SimplifiedAssets.zip](https://github.com/o3de/o3de/files/9500395/SimplifiedAssets.zip)

**Comment by chiyenteng** [comment]
Related GHI: https://github.com/o3de/o3de/pull/10730

**Comment by chiyenteng** [comment]
Although I couldn't repro the crash on my machine, @forkercat is still seeing crashes consistently on his machine:
![image (1)](https://user-images.githubusercontent.com/82238204/189704858-2a838efa-fe79-4d91-9fa3-be0b63472a80.png)
![image (2)](https://user-images.githubusercontent.com/82238204/189704981-eb2ce82a-4225-43c0-96d0-dee94cc7ac7e.png)

Adding similar logic to endIndex but still can't resolve the issue:
![image (3)](https://user-images.githubusercontent.com/82238204/189705257-65ee404a-2099-440b-b38d-a3d79da73281.png)

**Comment by chiyenteng** [comment]
Might need someone who can repro this issue to look into function `EntityOutlinerTreeView::dataChanged` to find actual cause of crashes and more details.

**Comment by michabr** [comment]
The crash happens during instance propagation when entities are being removed from the outliner to be replaced with new ones.

**Comment by forkercat** [comment]
## Update 9/28

I am still able to repro it in recent dev/stab with ***SimplifiedAssets.zip*** I provided above.

https://user-images.githubusercontent.com/11157226/192887901-4230bd8c-9e81-42f1-aaae-41cc361bfe44.mp4

I entered and exited the focus mode by clicking the icon. The editor freezes after couple attempts.

**Comment by forkercat** [comment]
I did some verifications that might be helpful.

***1. It only crashes on this particular hierarchy with 3 nested prefabs.***

```cpp
Level
  TruckPrefab    (prefab)
    Truck
      WheelsPrefab    (prefab)
        Wheels
          WheelPrefab    (prefab)
            Wheel
```

Other hierarchies like the one below are okay.

```cpp
Level
  TruckPrefab    (prefab)
    Truck
    WheelsPrefab    (prefab)
      Wheels
      WheelPrefab    (prefab)
        Wheel
```

***2. It did not crash before, so it could be related to some changes we introduced recently.***

These are some related PRs in my mind:
- July 19 - @AMZN-daimini put up #10730 to fix a similar issue. I verified that the crash cannot be reproduced on this commit.
- Aug 9 - @forkercat put up #11120 to fix other kind of crash. I verified that the crash cannot be reproduced on this commit.

**Comment by forkercat** [comment]
## Update 9/30

- I was trying to identify when the crash started happening. From the list above, the crash can be reproduced on and after this PR. I verified that one commit before does not crash. Now, I am looking into the commit to see which change is related. Tagging @chiyenteng @AMZN-daimini for visibility
    - https://github.com/o3de/o3de/pull/10943
- @alexmontAmazon can also repro it and is looking into Qt call stack to see if we can get more info about the crash.

**Comment by tonybalandiuk** [comment]
thank you @forkercat  please provide another update at end of day 9/30 if possible.

**Comment by forkercat** [comment]
**EOD Update:** PR is up to fix and it is pending for review and passing AR. Once it is green, it will be merged next Monday 10/3.

**Comment by tonybalandiuk** [comment]
thanks for the update. Note @monroegm (Sig-Content) that once this PR is prepared, it will require an exception which should be easy to get.

**Comment by jckand-amzn** [comment]
Verified fix on stabilization2210@ 7c4e3ca2689b78c872acf23fe5339b4b45dddf5f.
